packages/excalidraw/data/library.ts

```diff
@@ -463,23 +546,135 @@ export const useHandleLibrary = ({
     };
 
     // -------------------------------------------------------------------------
-    // ------ init load --------------------------------------------------------
-    if (getInitialLibraryRef.current) {
-      excalidrawAPI.updateLibrary({
-        libraryItems: getInitialLibraryRef.current(),
-      });
-    }
+    // ---------------------------------- init ---------------------------------
+    // -------------------------------------------------------------------------
 
     const libraryUrlTokens = parseLibraryTokensFromUrl();
 
     if (libraryUrlTokens) {
       importLibraryFromURL(libraryUrlTokens);
     }
+
+    // ------ (A) init load (legacy) -------------------------------------------
+    if (
+      "getInitialLibraryItems" in optsRef.current &&
+      optsRef.current.getInitialLibraryItems
+    ) {
+      console.warn(
+        "useHandleLibrar `opts.getInitialLibraryItems` is deprecated. Use `opts.adapter` instead.",
+      );
+
+      Promise.resolve(optsRef.current.getInitialLibraryItems())
+        .then((libraryItems) => {
+          excalidrawAPI.updateLibrary({
+            libraryItems,
+            merge: true,
+          });
+        })
+        .catch((error: any) => {
+          console.error(
+            `UseHandeLibrary getInitialLibraryItems failed: ${error?.message}`,
+          );
+        });
+    }
+
+    // -------------------------------------------------------------------------
     // --------------------------------------------------------- init load -----
+    // -------------------------------------------------------------------------
+
+    // ------ (A) data source adapter ------------------------------------------
+    let unsubOnLibraryChange: () => void | undefined;
+
+    if ("adapter" in optsRef.current && optsRef.current.adapter) {
+      const adapter = optsRef.current.adapter;
+
+      const persistLibraryChange = async (change: LibraryChange) => {
+        const IDBData = await adapter.load();
+
+        const nextLibraryItemsMap = arrayToMap(IDBData || []);
+
+        for (const [id] of change.deleted) {
+          nextLibraryItemsMap.delete(id);
+        }
+
+        const addedItems: LibraryItem[] = [];
+
+        for (const [id, item] of change.inserted) {
+          if (nextLibraryItemsMap.has(id)) {
+            // replace item with latest version
+            nextLibraryItemsMap.set(id, item);
+          } else {
+            addedItems.push(item);
+          }
+        }
+
+        const nextLibraryItems = addedItems.concat(
+          Array.from(nextLibraryItemsMap.values()),
+        );
+
+        await adapter.save(nextLibraryItems);
+
+        return nextLibraryItems;
+      };
+
+      const initDataPromise = resolvablePromise<LibraryPersistedData | null>();
+
+      // migrate from old data source if needed
+      // (note that we don't queue this operation as the saves are commutative)
```

Comment:
by "saves are commutative" I mean that since now we're only adding/removing items based on changes, it doesn't matter what order we save in. (At least it shouldn't.)

Reply:
not sure I follow here completely what do to queue here and why is it related to migrations, but set of added / removed elements is not commutative if it touches the same element - it makes a difference if you first delete the element and then (re)insert it, or do it the other way around - in first case you end up with one element, in second you have none

Reply:
I've changed the comment to explain that it's about insertion-only, item-scope level only, which we care about mostly during init.